### PR TITLE
Deformable Convolution v2 cpp extension

### PR DESCRIPTION
**Pull request by yhcao6**
The performance is as follow (use default lr and batch size, reported on v100):

| Model                     | Mem (GB) | Train time (s/iter) | box AP | mask AP |
|---------------------------|----------|---------------------|--------|---------|
| Faster-R-50-FPN-C3        | 4.5      | 0.623               | 40.2   |         |
| Faster-R-50-FPN-C3-mod    | 4.8      | 0.628               | 40.3   |         |
| Faster-R-50-FPN-dpool-mod | 6.9      | 0.812               | 38.2   |         |
| Faster-R-101-FPN-C3       | 6.7      | 0.849               | 41.9   |         |
| Mask-R-50-FPN-C3          | 6.0      | 0.738               | 41.2   | 37.4    |
| Mask-R-50-FPN-C3-mod      | 5.8      | 0.748               | 41.5   | 37.5    |
| Mask-R-101-FPN-C3         | 8.2      | 0.849               | 42.8   | 38.5    |
| Cascade-R-50-FPN-C3       | 4.8      | 0.693               | 43.8   |         |
| Cascade-R-101-FPN-C3      | 7.6      | 0.915               | 44.9   |         |
| Cascade-Mask-R-50-FPN-C3  | 7.3      | 0.998               | 44.6   | 38.5    |
| Cascade-Mask-R-101-FPN-C3 | 9.7      | 1.194               | 45.8   | 39.6    |

### CODE REVIEW

File: mmdet/ops/dcn/functions/modulated_dcn_func.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+
+class ModulatedDeformConvFunction(Function):
+
+    def __init__(ctx, stride, padding, dilation=1, deformable_groups=1):
+        super(ModulatedDeformConvFunction, ctx).__init__()
+        ctx.stride = stride
+        ctx.padding = padding
+        ctx.dilation = dilation
+        ctx.deformable_groups = deformable_groups
```
Comment (hellock) [inline mmdet/ops/dcn/functions/modulated_dcn_func.py]: The `__init__` method is not necessary. Arguments like `stride`, `padding` can be set in the forward method.

File: mmdet/ops/dcn/functions/modulated_dcn_func.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
+#!/usr/bin/env python
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
Comment (hellock) [inline mmdet/ops/dcn/functions/modulated_dcn_func.py]: We support python 3 only, so these import statements can be removed.

File: mmdet/ops/dcn/functions/modulated_dcn_func.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        height, width = input.shape[2:4]
+        kernel_h, kernel_w = weight.shape[2:4]
+        height_out = (height + 2 * ctx.padding -
+                      (ctx.dilation * (kernel_h - 1) + 1)) // ctx.stride + 1
+        width_out = (width + 2 * ctx.padding -
+                     (ctx.dilation * (kernel_w - 1) + 1)) // ctx.stride + 1
+        return n, channels_out, height_out, width_out
+
+
+class DeformRoIPoolingFunction(Function):
```
Comment (hellock) [inline mmdet/ops/dcn/functions/modulated_dcn_func.py]: Deformable pooling functions can be moved to an isolated py file.

File: mmdet/ops/dcn/modules/deform_conv.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+import torch
+import torch.nn as nn
+from torch.nn.modules.module import Module
+from torch.nn.modules.utils import _pair
+
+from ..functions.deform_conv import deform_conv
+
+
+class DeformConv(Module):
```
Comment (hellock) [inline mmdet/ops/dcn/modules/deform_conv.py]: To be consistent with other modules, use `DeformConv(nn.Module)`, and `from torch.nn.modules.module import Module` is unnecessary.

File: mmdet/ops/dcn/modules/deform_conv.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 kernel_size,
+                 stride,
+                 padding,
+                 dilation=1,
+                 deformable_groups=1,
+                 no_bias=True):
```
Comment (hellock) [inline mmdet/ops/dcn/modules/deform_conv.py]: To be consistent with the normal Conv2d, use `bias=False` instead of `no_bias=True`.

File: mmdet/ops/dcn/functions/deform_pool.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+class DeformRoIPoolingFunction(Function):
+
+    @staticmethod
+    def forward(ctx,
+                data,
+                rois,
+                offset,
+                spatial_scale,
+                pooled_size,
```
Comment (hellock) [inline mmdet/ops/dcn/functions/deform_pool.py]: Make argument names consistent with RoIPooling and RoIAlign.

File: mmdet/ops/dcn/src/modulated_dcn_cuda.cpp [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+                                const int num_bbox,
+                                const int channels_trans,
+                                const int no_trans,
+                                const float spatial_scale,
+                                const int output_dim,
+                                const int group_size,
+                                const int pooled_size,
+                                const int part_size,
+                                const int sample_per_part,
+                                const float trans_std);
```
Comment (hellock) [inline mmdet/ops/dcn/src/modulated_dcn_cuda.cpp]: Split deform pool functions into another cpp file.

File: compile.sh [hunk 8]
```diff
@@ -20,3 +20,16 @@ echo "Building nms op..."
...
+make clean
+make PYTHON=${PYTHON}
+
+echo "Building dcn..."
+cd ../dcn
+if [ -d "build" ]; then
+    rm -r build
+fi
+$PYTHON setup.py build_ext --inplace
+$PYTHON setup_modulated.py build_ext --inplace
```
Comment (hellock) [inline compile.sh]: This line should be deleted.

File: mmdet/ops/dcn/functions/deform_conv.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    @staticmethod
+    def forward(ctx,
+                input,
+                offset,
+                weight,
+                stride=1,
+                padding=0,
+                dilation=1,
+                deformable_groups=1,
+                im2col_step=64):
```
Comment (hellock) [inline mmdet/ops/dcn/functions/deform_conv.py]: The `bias` argument can be also added for convenience. An assertion can be used to ensure it is False.

File: mmdet/ops/dcn/functions/deform_conv.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    def forward(ctx,
+                input,
+                offset,
+                mask,
+                weight,
+                bias,
+                stride,
+                padding,
+                dilation=1,
+                deformable_groups=1):
```
Comment (hellock) [inline mmdet/ops/dcn/functions/deform_conv.py]: The argument list is not consistent with `DeformConvFunction`, which should follow `torch.nn.functional.conv2d`.

File: mmdet/ops/dcn/functions/deform_conv.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+        ctx.stride = _pair(stride)
+        ctx.padding = _pair(padding)
+        ctx.dilation = _pair(dilation)
+        ctx.deformable_groups = deformable_groups
+        ctx.im2col_step = im2col_step
+
+        ctx.save_for_backward(input, offset, weight)
+
+        output = input.new(*DeformConvFunction._output_size(
+            input, weight, ctx.padding, ctx.dilation, ctx.stride))
```
Comment (hellock) [inline mmdet/ops/dcn/functions/deform_conv.py]: use `new_empty()` so that a tuple can be used as the argument.
